### Perspective API Exploration

First, we have a dataset of Wikipedia comments made available by Jigsaw, a subsidiary of Google that created the Perspective tool. The dataset includes a unique comment id, the text of the comment, and a series of binary labels applied by human raters: "toxic," "severe_toxic," "obscene," "threat," "insult," and "identity_hate." I have appended the "score" column, which represents the toxicity score assigned to the comment text by the live version of the Perspective API. The data is available under a CC0 license.

In [1]:
import pandas as pd
import time

df = pd.read_csv('/Users/melaniehuynh/Downloads/labeled_and_scored_comments.csv')

In [2]:
df.sort_values(['score'])

id                                       comment_text  \
1158   61a85266de70cca0                         Можешь говорить по русски.   
5246   030a9ba0b32b4456  Rodzina Krusenstern Potomkowie Admirała Adama ...   
11040  a6814193ac6b4705  The opening paragraph states that the game is ...   
24192  f0d6dee96e057a5b  The book is used as a source in a number of ar...   
2332   cbb9268d3c7429c1     Vals Vienés \n |Rumba \n |Cha Cha Chá(ciclo 7)   
...                 ...                                                ...   
19994  1429a2c930688317                          you fucking piece of shit   
28728  d82e45eb103ad07b                   fuck off you stupid aspy asshole   
25535  e9b92c6c6b01aeef  FUCK YOU YOU PIECE OF SHIT \n FUCK YOU YOU PIE...   
25945  e783fd267f3a9d3b  FUCK WIKIPEDIA ON WHEELS! \n\nFuck off wikiped...   
17833  364b6c24210e2345  Fuck you you fucking pig!\n\nYou motherfucking...   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  
1158   0.000004      0             0        0       0       0              0  
5246   0.000009      0             0        0       0       0              0  
11040  0.000091      0             0        0       0       0              0  
24192  0.000116      0             0        0       0       0              0  
2332   0.000131      0             0        0       0       0              0  
...         ...    ...           ...      ...     ...     ...            ...  
19994  0.997278      1             1        1       0       1              0  
28728  0.997736      1             1        1       0       1              0  
25535  0.997982      1             1        1       0       1              0  
25945  0.998136      1             1        1       0       1              0  
17833  0.998329      1             0        1       0       1              0  

[41338 rows x 9 columns]

# Data Parsing
After reading in the data, I imported the Counter subclass to get the count of the most common (top 50) words in the dataset.

In [3]:
from collections import Counter
Counter(" ".join(df["comment_text"]).split()).most_common(50)

[('the', 115302),
 ('to', 74441),
 ('of', 56293),
 ('and', 53227),
 ('a', 52575),
 ('I', 47409),
 ('is', 42646),
 ('you', 40143),
 ('that', 36143),
 ('in', 33849),
 ('for', 24438),
 ('it', 23903),
 ('on', 21678),
 ('"', 21308),
 ('not', 21165),
 ('be', 20857),
 ('this', 18404),
 ('have', 17832),
 ('as', 17810),
 ('are', 17031),
 ('your', 14407),
 ('with', 14380),
 ('was', 13569),
 ('or', 12136),
 ('but', 10943),
 ('an', 10687),
 ('The', 10233),
 ('from', 10055),
 ('article', 9994),
 ('my', 9899),
 ('by', 9884),
 ('at', 9510),
 ('about', 8879),
 ('do', 8229),
 ('if', 8032),
 ('can', 7837),
 ('has', 7701),
 ('would', 7319),
 ('will', 7289),
 ('page', 6923),
 ('me', 6629),
 ('what', 6609),
 ('so', 6533),
 ('like', 6480),
 ('If', 6431),
 ('been', 6280),
 ('just', 6191),
 ('all', 6098),
 ('any', 6082),
 ('-', 6066)]

# Finding Toxicity

In order to get a sense of the most "toxic" words, I used a for loop to iterate through the comment_text column, and obtained the comments with a score of greater than 0.99.

In [4]:
sum_toxic = 0
for index, row in df.iterrows():
    #print(row['comment_text'])
    if df['score'][index] > 0.99:
        print(df['comment_text'][index])
        sum_toxic += 1

print (sum_toxic)

:::Jeez Ed, you seem like a fucking shitty douchebag
==Son of a bitch== 

 Hey you fucking neo-Nazi, murkin son of a bitch. FUck you. Get the fuck out of this country or move to US of fucking A you motherfucker.  

 P.S> 

 Don't let the door hit your ass on the fucking way out, you fucking bushite.
whats your go cunt deleting my shit
FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK YOU ALL! FUCK 

Similarly, in order to get a sense of the least "toxic" words, I used a for loop to iterate through the comment_text column, and obtained the comments with a score of less than 0.01.

In [5]:
sum_non_toxic = 0
for index, row in df.iterrows():
    #print(row['comment_text'])
    if df['score'][index] < 0.01:
        print(df['comment_text'][index])
        sum_non_toxic += 1

print (sum_non_toxic)

One example of this is a used Boeing 767-200 that was purchased by the founders of Google, Sergey Brin and Larry Page, and later converted.
* p 47: Not just oddballs: mathematician Simon Plouffe (coinventor of BPP algorithm) set memorization record in 1975 with 4,096 digits (broken 1977). 
 * p 47:  current record data (obsolete) 
 * (no page)  modern records:  Akira Haraguchi - 100,000 digits; Lu Chao  67,890  digits.
There are also articles on La Liga, the Copa del Rey, Copa America and the Copa Libertadores.
| date2 = 8 April 2015 | result2 = no consensus | page2 = Kaun Banega Champu (2nd nomination)
* May 2006 – June 20 2006
" 
 ::both the britannica article and the Jewish encyclopedia article state in their first sentence that the source of title ""the ten commandments"" is from the biblical עשרת הדברים.   "
aab tumhare havale vatan sathiyon
Можешь говорить по русски.
The pattern for inclusion could follow the example of the following set of four lists. 
 ::*Wikipedia:Vital articl

In [6]:
df['comment_text'].mode()

0        !\n\nIt is now inside your bones, your head. H...
1                    !  Go interfere with someone else!'''
2        ! IT WAS MENTIONED IN THE MOVIE THAT SHE WAS H...
3                                 ! NIGGERS IN THE HOUSE >
4        ! Please do not remove or change this AfD mess...
                               ...                        
41333    只, is it correctly written in the picture of t...
41334    第一課 冬休みが終わって \n\n 本课词汇 \n 久しぶり（ひさしぶり）（名　形動）⑤〇 ...
41335    訛古乃 Go-koò-naè, Dog with variegated hair.........
41336    这是什么意思？\n\nYou have been blocked from editing....
41337                         일이삼사오육칠팔구하고십이요 에헤헤 으헤 으헤 으허허
Name: comment_text, Length: 41338, dtype: object

# Using the API

In [7]:
from googleapiclient.discovery import build
import json

def get_toxicity_score(comment):
    
  API_KEY = 'XXXXX' # Put your API key here
    
  client = build(
  "commentanalyzer",
  "v1alpha1",
  developerKey=API_KEY,
  discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
  )

  analyze_request = {
  'comment': { 'text': comment },
  'requestedAttributes': {'TOXICITY': {}}
  }
    
  response = client.comments().analyze(body=analyze_request).execute()
  toxicity_score = response["attributeScores"]["TOXICITY"]["summaryScore"]["value"]
    
  return toxicity_score

In [8]:
comment_list = ['hello', 'how are you', 'fine thanks']
score_list = []

for comment in comment_list:
    score = get_toxicity_score(comment)
    score_list.append(score)
    time.sleep(1)

sum(score_list)

0.056184514

# Designing and Performing Tests

*(Profanity warning for testing purposes!)* - Here, I was interested in obtaining the toxicity scores for various gendered curse words in order to see if any bias could be traced back to misogynistic insults/swears.

The first set of words I wanted to run tests on was a random set of words to understand how the toxicity scores range.

The second set of words I wanted to test abusive language with a denotation/connotation tied to the female gender - words such as "b/tch", "wh/re, and sl/t" versus words tied to the male gender - "d/ck, assh/le, and douchebag."

The third set of words are miscellaneous offensive words that I utilized to gauge how offensive language was scored.

In [9]:
get_toxicity_score("hello")

0.02397547

In [10]:
get_toxicity_score("she")

0.030155903

In [11]:
get_toxicity_score("he")

0.02792408

In the above first set, it is interesting to see how the pronoun "she" has a higher toxicity score than "he".

In [12]:
get_toxicity_score("bitch")

0.9254919

In [13]:
get_toxicity_score("whore")

0.78524166

In [14]:
get_toxicity_score("slut")

0.60152835

In [15]:
get_toxicity_score("dick")

0.59863794

In [16]:
get_toxicity_score("asshole")

0.89158845

In [17]:
get_toxicity_score("douchebag")

0.53430706

In [18]:
get_toxicity_score("stupid")

0.7060645

In [19]:
get_toxicity_score("idiot")

0.8086813

In [20]:
get_toxicity_score("dumb")

0.7007861

After summing the scores of this set, it seems that Perspective denotes more toxicity to popular swear words relating to verbal aggression historically tied with femininity/women. 

###### Utilizing the toxicity scores above, I would determine the threshold for toxic or abusive words would be > 0.2 using the "Score" label.

Below, I tested the two sets of words associated with feminine versus masculine identities to see the sum of the scores and the number of mentions within the dataset.

In [21]:
comment_list = ['bitch', 'whore', 'slut']
fscore_list = []

for comment in comment_list:
    score = get_toxicity_score(comment)
    fscore_list.append(score)
    time.sleep(1)

sum(fscore_list)

2.31226191

In [22]:
comment_list = ['dick', 'asshole', 'douchebag']
mscore_list = []

for comment in comment_list:
    score = get_toxicity_score(comment)
    mscore_list.append(score)
    time.sleep(1)

sum(mscore_list)

2.02453345

In [23]:
fswear_count = df['comment_text'].str.contains('bitch').sum()
fswear_count += df['comment_text'].str.contains('whore').sum()
fswear_count += df['comment_text'].str.contains('slut').sum()
if fswear_count > 0:
    print ("There are {f} mentions of the words 'b*tch', 'wh*re', and 'sl*t.'".format(f = fswear_count))

There are 238 mentions of the words 'b*tch', 'wh*re', and 'sl*t.'


In [24]:
mswear_count = df['comment_text'].str.contains('dick').sum()
mswear_count += df['comment_text'].str.contains('asshole').sum()
mswear_count += df['comment_text'].str.contains('douchebag').sum()
if mswear_count > 0:
    print ("There are {m} mentions of the words 'd*ck', 'assh*le', and 'douchebag.'".format(m = mswear_count))

There are 341 mentions of the words 'd*ck', 'assh*le', and 'douchebag.'


# Testing the Hypothesis

##### Hypothesis: Perspective is more likely to mark anti-female gendered swears as toxic/the toxicity score will be higher when compared to non-gendered or anti-male swears (excluding f***).

Utilizing .loc and str.contains, I organized the data to sort through different sets of swear words (anti-female, anti-male, non-gendered, less explicit/offensive).

In [25]:
df_female_swears = df.loc[df.comment_text.str.contains(r'\b(?:bitch|whore|slut)\b')]
df_female_swears.describe()

score       toxic  severe_toxic     obscene      threat  \
count  179.000000  179.000000    179.000000  179.000000  179.000000   
mean     0.927538    0.865922      0.240223    0.837989    0.067039   
std      0.115469    0.341692      0.428417    0.369495    0.250791   
min      0.206598    0.000000      0.000000    0.000000    0.000000   
25%      0.932329    1.000000      0.000000    1.000000    0.000000   
50%      0.967966    1.000000      0.000000    1.000000    0.000000   
75%      0.988578    1.000000      0.000000    1.000000    0.000000   
max      0.996872    1.000000      1.000000    1.000000    1.000000   

          insult  identity_hate  
count  179.00000     179.000000  
mean     0.77095       0.173184  
std      0.42140       0.379468  
min      0.00000       0.000000  
25%      1.00000       0.000000  
50%      1.00000       0.000000  
75%      1.00000       0.000000  
max      1.00000       1.000000

In [26]:
df_male_swears = df.loc[df.comment_text.str.contains(r'\b(?:dick|asshole|douchebag)\b')]
df_male_swears.describe()

score       toxic  severe_toxic     obscene      threat  \
count  265.000000  265.000000    265.000000  265.000000  265.000000   
mean     0.918081    0.837736      0.200000    0.803774    0.041509   
std      0.099548    0.369391      0.400757    0.397893    0.199843   
min      0.427179    0.000000      0.000000    0.000000    0.000000   
25%      0.896489    1.000000      0.000000    1.000000    0.000000   
50%      0.949928    1.000000      0.000000    1.000000    0.000000   
75%      0.980579    1.000000      0.000000    1.000000    0.000000   
max      0.997736    1.000000      1.000000    1.000000    1.000000   

           insult  identity_hate  
count  265.000000     265.000000  
mean     0.694340       0.105660  
std      0.461558       0.307984  
min      0.000000       0.000000  
25%      0.000000       0.000000  
50%      1.000000       0.000000  
75%      1.000000       0.000000  
max      1.000000       1.000000

In [27]:
df_other_swears = df.loc[df.comment_text.str.contains(r'\b(?:fuck|fucking|motherfucking)\b')]
df_other_swears.describe()

score       toxic  severe_toxic     obscene      threat  \
count  732.000000  732.000000    732.000000  732.000000  732.000000   
mean     0.920826    0.870219      0.206284    0.842896    0.040984   
std      0.097565    0.336293      0.404914    0.364147    0.198388   
min      0.271009    0.000000      0.000000    0.000000    0.000000   
25%      0.898114    1.000000      0.000000    1.000000    0.000000   
50%      0.958911    1.000000      0.000000    1.000000    0.000000   
75%      0.981523    1.000000      0.000000    1.000000    0.000000   
max      0.998329    1.000000      1.000000    1.000000    1.000000   

           insult  identity_hate  
count  732.000000     732.000000  
mean     0.635246       0.113388  
std      0.481690       0.317283  
min      0.000000       0.000000  
25%      0.000000       0.000000  
50%      1.000000       0.000000  
75%      1.000000       0.000000  
max      1.000000       1.000000

In [28]:
df_other_swears_2 = df.loc[df.comment_text.str.contains(r'\b(?:stupid|idiot|dumb)\b')]
df_other_swears_2.describe()

score       toxic  severe_toxic     obscene      threat  \
count  681.000000  681.000000    681.000000  681.000000  681.000000   
mean     0.801092    0.581498      0.036711    0.281938    0.010279   
std      0.162395    0.493676      0.188189    0.450274    0.100937   
min      0.153055    0.000000      0.000000    0.000000    0.000000   
25%      0.702893    0.000000      0.000000    0.000000    0.000000   
50%      0.837820    1.000000      0.000000    0.000000    0.000000   
75%      0.932329    1.000000      0.000000    1.000000    0.000000   
max      0.997736    1.000000      1.000000    1.000000    1.000000   

           insult  identity_hate  
count  681.000000     681.000000  
mean     0.418502       0.042584  
std      0.493676       0.202067  
min      0.000000       0.000000  
25%      0.000000       0.000000  
50%      0.000000       0.000000  
75%      1.000000       0.000000  
max      1.000000       1.000000

# Analyzing
When assessing these numbers, the set of anti-female gendered swears has the highest score mean (0.9257). The anti-male gendered swears has a mean of 0.918, non-gendered of 0.920, and other offensive/less explicit words has a mean of 0.801. This data, so far, aligns with my hypothesis - Perspective seems to mark anti-female terms as more offensive than anti-male/non-gendered terms.

In [29]:
get_toxicity_score("Stay in the kitchen where you belong")

0.45811868

However, there are errors with this method of testing - there are many other curse words that may have anti-female contexts, and there are phrases that are misogynistic (above) that are not included when just looking for single offensive words. The above phrase is a notably anti-female statement; and yet, the toxicity score is much lower than with the use of a single offensive word.

In [30]:
df_female_swears.sort_values(['score'])

id                                       comment_text  \
10951  a80b56df3dcd3b7f  "\n\n Trivia lesson  \n\nOk, I’m going to enjo...   
6286   f34b28e44355924f  "\n\nI copy this from comfort women pages. It ...   
31705  c85b7946132c29da  "== Poster ==\n\nDoes the poster for this arti...   
19078  229cc6cdff7035e4  "For those who want to bitch about me any beha...   
2903   917ad275221a1329  == LOCK THIS PAGE so that only registered user...   
...                 ...                                                ...   
21751  fd6e58a896118b31                      fuck off you son of a bitch!!   
23096  f68d9ef333909f03  Fucking whore \n\nYou are such a dumb bitch i ...   
35749  b25c87ee437e7e51  why? \n\nfuck you you fucking whore.\nyou stup...   
18792  26fed2949c2dadc1            you son of a bitch RIP you motherfucker   
6964   e88321c89a619419     Stupid fucking bitch rot in hell piece of shit   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  
10951  0.206598      0             0        0       0       0              0  
6286   0.304305      0             0        0       0       0              0  
31705  0.366981      0             0        0       0       0              0  
19078  0.464582      0             0        0       0       0              0  
2903   0.628260      0             0        0       0       0              0  
...         ...    ...           ...      ...     ...     ...            ...  
21751  0.994951      1             1        1       0       1              0  
23096  0.995611      1             0        1       0       1              0  
35749  0.996048      1             1        1       0       1              0  
18792  0.996669      1             0        1       0       1              0  
6964   0.996872      1             1        1       0       1              0  

[179 rows x 9 columns]

In [31]:
df_male_swears.sort_values(['score'])

id                                       comment_text  \
37340  a9c317bd7f49e7bc  "\n\n Well, in a perfect world it would depend...   
10580  ae27f949bbd5be02  "{| class=""messagebox standard-talk"" id=""ta...   
27120  e1226e5c80278184  Hey. I meant to thank you for your stance, but...   
10315  b204c21f06e3219f  Lofat \n\nHi Saposcat,\n\n adviced to ask you....   
14384  6f51296b928b4bf4  "\n\n re ""I thought he was kind of a dick yes...   
...                 ...                                                ...   
15413  5e4fb1710d0f1493  your moms yummy cunt \n\nFucking Bastard\nyou ...   
19899  15be2045a73859f5  You're a motherfucking prick! \n\nSo, you don'...   
17266  3f640e15e1903bea  You are a mother fucking son of a bich you ass...   
28164  db44878f613cc516              Youre an asshole \n\nyoure an asshole   
28728  d82e45eb103ad07b                   fuck off you stupid aspy asshole   

          score  toxic  severe_toxic  obscene  threat  insult  identity_hate  
37340  0.427179      0             0        0       0       0              0  
10580  0.467896      0             0        0       0       0              0  
27120  0.530155      0             0        0       0       0              0  
10315  0.534903      0             0        0       0       0              0  
14384  0.534903      0             0        0       0       0              0  
...         ...    ...           ...      ...     ...     ...            ...  
15413  0.993966      1             1        1       0       1              0  
19899  0.994153      1             0        1       1       1              0  
17266  0.994164      1             0        1       0       1              0  
28164  0.994474      1             0        1       0       1              0  
28728  0.997736      1             1        1       0       1              0  

[265 rows x 9 columns]

In [32]:
df_female_swears[(df_female_swears['toxic']==1)|(df_female_swears['severe_toxic']==1)].head()

id                                       comment_text  \
62    03c6344729c4e665  Azerbaijanis Proud of Ramil Safarov  \n If I h...   
332   1a44cad92a0c1c96  ==Son of a bitch== \n\n Hey you fucking neo-Na...   
476   2433d7112e3da3c6  " \n\n  \n\n :::I have heard people describe M...   
836   4368d67cfeb54881  Deleted 3-2  \n You a bitch ass nigga for dele...   
1027  565d151901ac2063  == Laura Bush Article == \n\n What the fuck di...   

         score  toxic  severe_toxic  obscene  threat  insult  identity_hate  
62    0.980994      1             0        1       1       1              1  
332   0.992705      1             1        1       0       1              1  
476   0.948086      1             0        1       0       1              0  
836   0.986299      1             1        1       0       1              1  
1027  0.989706      1             1        1       0       1              0

In [33]:
threshold = 0.2

df_female_swears['prediction'] = (df_female_swears['score'] > threshold).astype(int)
df_female_swears['prediction'].value_counts()

/var/folders/8r/jdqm68r537s7vc8l4h_74z9w0000gn/T/ipykernel_5099/1396816425.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_female_swears['prediction'] = (df_female_swears['score'] > threshold).astype(int)


1    179
Name: prediction, dtype: int64

In [34]:
threshold = 0.2

df_male_swears['prediction'] = (df_male_swears['score'] > threshold).astype(int)
df_male_swears['prediction'].value_counts()

/var/folders/8r/jdqm68r537s7vc8l4h_74z9w0000gn/T/ipykernel_5099/3547238737.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_male_swears['prediction'] = (df_male_swears['score'] > threshold).astype(int)


1    265
Name: prediction, dtype: int64

In [35]:
threshold = 0.2

df_other_swears['prediction'] = (df_other_swears['score'] > threshold).astype(int)
df_other_swears['prediction'].value_counts()

/var/folders/8r/jdqm68r537s7vc8l4h_74z9w0000gn/T/ipykernel_5099/689905649.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_other_swears['prediction'] = (df_other_swears['score'] > threshold).astype(int)


1    732
Name: prediction, dtype: int64

In [36]:
threshold = 0.2

df_other_swears_2['prediction'] = (df_other_swears_2['score'] > threshold).astype(int)
df_other_swears_2['prediction'].value_counts()

/var/folders/8r/jdqm68r537s7vc8l4h_74z9w0000gn/T/ipykernel_5099/3618820.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_other_swears_2['prediction'] = (df_other_swears_2['score'] > threshold).astype(int)


1    679
0      2
Name: prediction, dtype: int64

In [37]:
from sklearn.metrics import confusion_matrix

In [38]:
confusion_matrix(df_female_swears['toxic'], df_female_swears['prediction'])

array([[  0,  24],
       [  0, 155]])

In [39]:
confusion_matrix(df_male_swears['toxic'], df_male_swears['prediction'])

array([[  0,  43],
       [  0, 222]])

In [40]:
confusion_matrix(df_other_swears['toxic'], df_other_swears['prediction'])

array([[  0,  95],
       [  0, 637]])

In [41]:
confusion_matrix(df_other_swears_2['toxic'], df_other_swears_2['prediction'])

array([[  2, 283],
       [  0, 396]])

# More Analysis
The amount of "toxic" anti-female swears compares 24 'toxic' words under the label to 155 predicted. (545.833% increase in prediction compared to actual)

The amount of "toxic" anti-male swears compares 43 'toxic' words under the label to 222 predicted. (416.279% increase in prediction compared to actual)

The amount of "toxic" non-gendered/other swears compares 95 'toxic' words under the label to 637 predicted. (570.526% increase in prediction compared to actual)

The amount of "toxic" non-gendered/less offensive swears compares 283 'toxic' words under the label to 396 predicted. (39.9293% increase in prediction compared to actual) 

The anti-female and anti-male swears, as well as the swears including f***, were largely overpredicted. Though not a fullproof analysis, it seems Perspective is more likely to mark anti-female remarks/swear words as "toxic," compared to anti-male or non-gendered swear words. The exception seems to be when anti-female remarks are stated within phrases instead of single offensive words.  